### ISSUE TITLE
Everest leaks file descriptors on Linux

### ISSUE THREAD

**Issue by ddevault** [EverestAPI/Everest#436]
### My mods and Everest install are up to date

Yes

### I have recreated the bug with only Everest OR a minimum number of mods enabled

No

### Describe the bug

Everest hits ENFILE on Linux because it has too many open file descriptors. Should double check that files are closed when no longer in use.

### Steps to reproduce

1. Install a lot of mods.
2. Poof

### Expected behavior

Should close files when done with them to avoid hitting the open file limit.

### Operating System

Linux

### Everest Version

1.4.0.0

### Mods required to reproduce

Any mods: just install enough of them to exhaust the fds.

**Comment by WEGFan** [comment]
Did you unzipped the mods? Everest will automatically watch the files in unzipped mods for the hot reload feature, I don't know if it's the issue. But if you aren't making mods you shouldn't unzip them.

**Comment by WEGFan** [comment]
Did you unzipped the mods? Everest will automatically watch the files in unzipped mods for the hot reload feature, I don't know if it's the issue. But if you aren't making mods you shouldn't unzip them.

**Comment by ddevault** [comment]
I did not unzip them, no.

**Comment by microlith57** [comment]
how many mods is enough for you, and what is your `ulimit -n`? on my system i have it set to > 1 million, so i don't think it's possible for me to reproduce this.

**Comment by ddevault** [comment]
I have 79 mods installed and the default ulimit of 1024. You can set a lower ulimit for testing with `ulimit -n <whatever>`.

**Comment by 0x0ade** [comment]
Does setting the environment variable `MONO_MANAGED_WATCHER` to `disabled` help in any way? I've heard that macOS has got a "low" ulimit as well (at least compared to other environments Everest is being used in), which makes me wish I had the time 
to move away from the C# standard `FileSystemWatcher` implementation. Maybe someone else would be willing to look into that too.

**Comment by ddevault** [comment]
Yep, setting that variable seems to prevent the crash.

**Comment by LeSeulArtichaut** [comment]
I ran into this bug while trying to play the [Strawberry Jam Collab](https://gamebanana.com/mods/424541). Setting `MONO_MANAGED_WATCHER=disabled` didn't seem to help, however setting a high `ulimit -n` did prevent the crash.

**Comment by Kalobi** [comment]
The current recommended solution is to disable FastTextureLoading in modsettings-Everest.celeste until we figure out how to limit the open files for FTL.

**Comment by nlburgin** [comment]
Was this ever fixed?

I'm still using a launcher script for Everest that works around this by calling  `ulimit -Sn 99999` before actually starting Everest. Is that still necessary?

**Comment by SnipUndercover** [comment]
I haven't seen a file descriptor limit crash in years at this point, about when Everest Core became live I believe. I think it's safe to close.

A new FTL implementation PR is around the corner as well, but I'm not sure whether it limits the open file count. @Wartori54

**Comment by nlburgin** [comment]
I think I was still on classic Mono instead of Core when I started using that workaround, so that makes some sense. I'll have to test loading Strawberry Jam without it later.

**Comment by SnipUndercover** [comment]
It may have been the case that every OS just has `ulimit -n` set to `unlimited` by default nowadays, I don't know. I do remember however a race condition making some mod zips permanently "in use", preventing them from being updated, but that shouldn't be enough to crash.

**Comment by nlburgin** [comment]
> It may have been the case that every OS just has `ulimit -n` set to `unlimited` by default nowadays, I don't know.

Default `ulimit -n` without explicitly declaring a higher one is still 1024 on Ubuntu family. Not sure _why_ it's trying to make sure that FD leakers will crash, but the limit is still quite low by default. 

This is an arbitrary soft limit and very easy to override, so I'm not really sure why it's there at all.

**Comment by SnipUndercover** [comment]
In any case, can you reproduce this without overriding the soft limit?

**Comment by nlburgin** [comment]
I commented out the ulimit line in my launcher script, and enabled Strawberry Jam with all its dependencies, which was the classic example of a mod pack big enough to trigger a startup crash with this bug. It didn't crash.

Guess it's fixed?

**Comment by SnipUndercover** [comment]
It could be that replacing `DotNetZip` with `System.IO.Compression` (#862, reverted and rereverted in c2341622242b9c3a628167deb09cbf92e7de48d1) fixed the file descriptor issues, that's the only possible way we could think of.

**Comment by Wartori54** [comment]
> I haven't seen a file descriptor limit crash in years at this point, about when Everest Core became live I believe. I think it's safe to close.
> 
> A new FTL implementation PR is around the corner as well, but I'm not sure whether it limits the open file count. @Wartori54 

It does not, should be around the same as the old one (mayhaps a little bit worse though).
I also have a patch for it that is not yet ready for PR that should close fds earlier, which could help.
Regardless I also believe the IO.Compression migration is what helped the most here.

**Comment by nlburgin** [comment]
Yeah, using a different library to unzip mods could have something to do with preventing the startup crash.

I do remember the behavior was always weird. Before I started using the ulimit workaround I remember it was still _possible_ to start strawberry jam, but only by not having all mods enabled at startup, so that Everest has to do its own dependency resolution and a load a portion of the mods later on _after_ startup.

This suggests that the explosion of FDs during startup wasn't quite a leak, just an excessive _delay_ in closing them...

I also remember that before increasing my ulimit, there was another weird crash besides the one at startup. One specific level, "Attack of the Clone" in Advanced, would always crash a couple rooms in. It had to also be related to an FD leak since the ulimit workaround fixed it along with the startup crash. I wonder if that relates to the other issue you mentioned, where the "Fast Texture Loading" also leaks FDs. I should probably test that again as well...

**Comment by nlburgin** [comment]
Just full-cleared "Attack of the Clone" again and it doesn't crash either. All the old crashes I needed to prevent by boosting the open-files limit, don't seem applicable anymore.

**Comment by SnipUndercover** [comment]
I suppose we can close this. If this shows up then they can make a new issue or reopen this one, I guess.
